### ISSUE TITLE
fix the MySQL driver

### ISSUE THREAD

**Issue by gavinking** [hibernate/hibernate-rx#37]
We need to try this stuff out on MySQL.

**Comment by gavinking** [comment]
Alright, I tried it out and the results were .... *disappointing*. The driver from MySQL has more or at least different bugs compared to the Postgres driver:

- First, unlike in JDBC, the drivers don't abstract over the database's native prepared statement parameter format. The `$1`, `$2`, etc we had to use for Postgres are *because that's what's native on PostgresSQL*! And so this syntax doesn't work in MySQL which expects `?`s like in JDBC. 🤦
- Second, the driver can't handle columns of MySQL type `BIT` in `SELECT` lists, where by "can't handle", I mean it just silently returns no rows. Hibernate uses `BIT` as the column type for `boolean` on MySQL. This was *very* hard to track down.
- Third, the driver doesn't implement its own `getBigDecimal()` method, which we use to read `NUMERIC`/`DECIMAL` columns. At least it throws an `UnsupportedOperationException` in this case.
- Nor does it implement its `getLocalTime()` method

Finally, my code isn't correctly interpreting `SequenceStyleGenerator`, which can apparently mean table generation on databases without sequences.

**Comment by gavinking** [comment]
Given the status of the driver, I would say that MySQL support is *not* a priority for the first release.

**Comment by BillyYccc** [comment]
Hi Gavin, I’m the author of Vert.x MySQL and MSSQL client and I assume you’re using this :-)

I would like to clarify a bit for the questions you mentioned above.

* First Vert.x SQL Client aims not to be an abstraction, which means it won’t parse the SQL text and unify the placeholders. This does not behave like what JDBC does and it would be good for users from backgrounds of different database(MySQL uses `?` and SQL Server uses `@P0`, `@P1`) and languages(JavaScript), this is more natural and less error-prone in my mind.
* MySQL `BIT` data type has been supported in the codebase but missed the latest release. It’s currently stored internally as a `java.lang.Long` type, and you can use `Row#getBoolean` accessor method to get a `Java.lang.Boolean` value, this also applies for `TINYINT` data type, it’s stored as a `java.lang.Byte` but can be retrieved as a boolean by the same `Row#getBoolean` method.
* `getBigDecimal` method is missing and it should be added.
* MySQL `TIME` is mapped to `java.time.Duration` instead of `java.time.LocalTime` because MySQL `TIME` value ranges from '-838:59:59' to '838:59:59' which behaves differently from other database servers.
* For the last point do you mean you can’t get the right last_insert_id() generated by MySQL?

**Comment by gavinking** [comment]
> Hi Gavin, I’m the author of Vert.x MySQL and MSSQL client and I assume you’re using this :-)

I'm using the `.axle` one, to be clear.

> First Vert.x SQL Client aims not to be an abstraction, which means it won’t parse the SQL text and unify the placeholders. This does not behave like what JDBC does and it would be good for users from backgrounds of different database(MySQL uses `?` and SQL Server uses `@P0`, `@P1`) and languages(JavaScript), this is more natural and less error-prone in my mind.

OK, but it's *extremely* surprising to Java or .Net developers who've only ever used JDBC/ODBC, and probably don't even *know* the syntax of their database's `prepare` statement. It's also a hurdle for anyone who's trying to make an existing Java lib work on this stuff, or is trying to write cross-platform code. There are really good reasons why ODBC and JDBC have this abstraction.

> MySQL BIT data type has been supported in the codebase but missed the latest release.

OK, good, thanks.

> `getBigDecimal` method is missing and it should be added.

Great.

> MySQL `TIME` is mapped to `java.time.Duration` instead of `java.time.LocalTime` because MySQL `TIME` value ranges from '-838:59:59' to '838:59:59' which behaves differently from other database servers.

OK, but:

1. `Duration`s aren't times. They're more similar to a SQL `interval` type. These things just don't have the same semantics at all.
2. if you have a `getLocalTime()` method on your API, it should do something useful.

Note: I don't mind if you do reasonable type conversions in the `getXxxxx()` methods, just as all JDBC drivers do.

> For the last point do you mean you can’t get the right last_insert_id() generated by MySQL?

No I just mean there was a bug in my (hibernate-rx) code. Don't worry about this one, we still have to do a bit of work before we can get identity column support working.

**Comment by gavinking** [comment]
By the way, there's another issue we ran into which is that when doing a result set lookup by column alias, the column aliases were always in lowercase, even if they were upper or mixed case in the SQL statement. We worked around this one already.

Thanks for your quick response, @BillyYccc, I'm not even sure how you found this.

**Comment by gavinking** [comment]
I've pushed the work I've already done on integrating MySQL, though some of it depends on #36.

**Comment by gavinking** [comment]
Well, I've got MySQL working well-enough that I can add a test. Note that properties of type `boolean`, `BigDecimal`, `BigInteger`, and `LocalTime` don't work in the current release of the driver.

My fixes depend on #36.

I've opened issues against SmallRye, so I can close this one now.
